### ISSUE TITLE
Server packet out of order when Prepared is destroyed too early

### ISSUE THREAD

**Issue by CromFr** [mysql-d/mysql-native#139]
```d
import mysql;

void main()
{
	auto conn = new Connection("xxx");
	scope(exit) conn.close();
	{// Works
		ResultRange result;

		auto prep = conn.prepare("SELECT ?");
		prep.setArgs("Hello world");
		result = prep.query();

		result.close();
	}
	{// Throws Server packet out of order
		ResultRange result;
		{
			auto prep = conn.prepare("SELECT ?");
			prep.setArgs("Hello world");
			result = prep.query();
		}

		result.close();
	}
}
```

From what I've understood, this happens every time the Prepared struct is destroyed while there is pending ResultRange results.

Note that if the Out of order exception is ignored, any following queries using the same connection will throw `Data is pending on the connection. Any existing ResultRange must be consumed or purged before performing any other communication with the server.`. This is a serious problem when using MySQLPool.

**Comment by Abscissa** [comment]
Ouch.

The next thing I want to do with mysql-native is eliminate "user must either read all results or manually purge them, otherwise an exception is thrown" in favor of "the library automatically purges unused results when needed". (See discussion in #117)

At a glance, I'd say it's possible that may fix this issue as well. I'll have to keep it in mind. Either way, this one is definitely high-priority.

**Comment by Abscissa** [comment]
I think I spotted the problem: The `release` function for prepares statements forgot to call `Connection.enforceNothingPending` - the function that's *supposed* to be used before sending the server any command so it can throw `MYXDataPending` instead of borking the whole comms stream. (No doubt occured because of insufficient encapsulation of low-level comms code - a historical legacy I need to fix, partly via #144).

The fix for #117, which I'm working on now, will definitely solve this issue too. If any similar cases are lingering, #144 should root them out.

**Comment by CromFr** [comment]
thanks for making D even more awesome :)
